File: 0-1 matrix.py
from typing import List
from queue import Queue

class Solution:
    DIR = [0, 1, 0, -1, 0]
    
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        m, n = len(mat), len(mat[0])
        q = Queue()
        for r in range(m):
            for c in range(n):
                if mat[r][c] == 0:
                    q.put((r, c))
                else:
                    mat[r][c] = -1 # Marked as not processed yet!

        while not q.empty():
            r, c = q.get()
            for i in range(4):
                nr, nc = r + self.DIR[i], c + self.DIR[i+1]
                if nr < 0 or nr == m or nc < 0 or nc == n or mat[nr][nc] != -1:
                    continue
                mat[nr][nc] = mat[r][c] + 1
                q.put((nr, nc))
        return mat

if __name__ == '__main__':
    n, m = map(int, input().split())
    matrix = []
    for i in range(n):
        matrix.append(list(map(int, input().split())))
    obj = Solution()
    res = obj.updateMatrix(matrix)
    for row in res:
        print(*row)

File: 01 Matrix.java
import java.util.*;

class Solution {
    private final int[] DIR = {0, 1, 0, -1, 0};
    
    public int[][] updateMatrix(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        Queue<int[]> q = new LinkedList<>();
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (mat[r][c] == 0) {
                    q.offer(new int[]{r, c});
                } else {
                    mat[r][c] = -1; // Marked as not processed yet!
                }
            }
        }

        while (!q.isEmpty()) {
            int[] pos = q.poll();
            int r = pos[0], c = pos[1];
            for (int i = 0; i < 4; ++i) {
                int nr = r + DIR[i], nc = c + DIR[i+1];
                if (nr < 0 || nr == m || nc < 0 || nc == n || mat[nr][nc] != -1) {
                    continue;
                }
                mat[nr][nc] = mat[r][c] + 1;
                q.offer(new int[]{nr, nc});
            }
        }
        return mat;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        int[][] matrix = new int[n][m];
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                matrix[i][j] = sc.nextInt();
            }
        }
        Solution obj = new Solution();
        int[][] res = obj.updateMatrix(matrix);
        for (int[] row : res) {
            for (int i : row) {
                System.out.print(i + " ");
            }
            System.out.println();
        }
    }
}
